reporter.go

```diff
@@ -392,126 +343,131 @@ func (r *Reporter) diffFile(before, after *fspb.File) (string, error) {
+func (r *Reporter) Compare(before, after *fspb.Walk) (*Report, error) {
+	if err := r.sanityCheck(before, after); err != nil {
+		return nil, err
 	}
-	r.Counter.Add(1, metric)
-}
 
-// Compare runs through two Walks (before and after) with a given ReportConfig and shows the diffs.
-func (r *Reporter) Compare(out io.Writer) {
-	// Processing report.
-	output := map[action][]actionData{}
 	walkedBefore := map[string]*fspb.File{}
 	walkedAfter := map[string]*fspb.File{}
-
-	if r.before != nil {
-		for _, fbOrig := range r.before.File {
+	if before != nil {
+		for _, fbOrig := range before.File {
 			fb := *fbOrig
 			fb.Path = NormalizePath(fb.Path, fb.Info.IsDir)
 			walkedBefore[fb.Path] = &fb
 		}
 	}
-	for _, faOrig := range r.after.File {
+	for _, faOrig := range after.File {
 		fa := *faOrig
 		fa.Path = NormalizePath(fa.Path, fa.Info.IsDir)
 		walkedAfter[fa.Path] = &fa
 	}
 
+	counter := metrics.Counter{}
+	output := Report{
+		Counter:    &counter,
+		WalkBefore: before,
+		WalkAfter:  after,
+	}
+
 	for _, fb := range walkedBefore {
-		r.count("before-files")
+		counter.Add(1, "before-files")
 		if r.isIgnored(fb.Path) {
-			r.count("before-files-ignored")
+			counter.Add(1, "before-files-ignored")
 			continue
 		}
 		fa := walkedAfter[fb.Path]
 		if fa == nil {
-			r.count("before-files-removed")
-			output[actionDelete] = append(output[actionDelete], actionData{before: fb})
+			counter.Add(1, "before-files-removed")
+			output.Deleted = append(output.Deleted, ActionData{Before: fb})
 			continue
 		}
 		diff, err := r.diffFile(fb, fa)
 		if err != nil {
-			r.count("file-diff-error")
-			output[actionError] = append(output[actionError], actionData{
-				before: fb,
-				after:  fa,
-				diff:   diff,
-				err:    err,
+			counter.Add(1, "file-diff-error")
+			output.Errors = append(output.Errors, ActionData{
+				Before: fb,
+				After:  fa,
+				Diff:   diff,
+				Err:    err,
 			})
 		}
 		if diff != "" {
-			r.count("before-files-modified")
-			output[actionModify] = append(output[actionModify], actionData{
-				before: fb,
-				after:  fa,
-				diff:   diff,
+			counter.Add(1, "before-files-modified")
+			output.Modified = append(output.Modified, ActionData{
+				Before: fb,
+				After:  fa,
+				Diff:   diff,
 			})
 		}
 	}
 	for _, fa := range walkedAfter {
-		r.count("after-files")
+		counter.Add(1, "after-files")
 		if r.isIgnored(fa.Path) {
-			r.count("after-files-ignored")
+			counter.Add(1, "after-files-ignored")
 			continue
 		}
 		_, ok := walkedBefore[fa.Path]
 		if ok {
 			continue
 		}
-		r.count("after-files-created")
-		output[actionAdd] = append(output[actionAdd], actionData{after: fa})
+		counter.Add(1, "after-files-created")
+		output.Added = append(output.Added, ActionData{After: fa})
 	}
+	return &output, nil
+}
 
-	// Writing sorted output.
+// PrintDiffSummary prints the diffs found in a Report.
+func (r *Reporter) PrintDiffSummary(out io.Writer, report *Report) {
```

Comment:
It might also make tests slitly simpler because you won't need to create a new `Reporter` just to call its `PrintDiffSummary` - no sideffects are expected anyway.

Reply:
I kept it as a method because of `r.Verbose` and because in the future there could be some other config that changes its behaviour and that way we don't have to change the signature.

I'm not against making it a function, but I'd either remove the verbose parameter (in this case it's not really that useful) or have all the `Print*` functions have it even if they don't use it, so that they are consistent.

Reply:
Sounds good. Can ways convert these to functions in the future in separate PRs if we wanted to.